### PR TITLE
USB reliability

### PR DESCRIPTION
**Pull request by laloch**
Improves USB storage reliability in case of concurrent access and/or slow media.

### CODE REVIEW

File: lib/WUI/nhttp/gcode_upload.cpp [hunk 1]
```diff
@@ -235,6 +264,7 @@ UploadHooks::Result GcodeUpload::finish(const char *final_filename, bool start_p
     snprintf(fname, sizeof fname, UPLOAD_TEMPLATE, file_idx);
 
     // Close the file first, otherwise it can't be moved
+    ftruncate(fileno(tmp_upload_file.get()), ftell(tmp_upload_file.get()));
```
Comment (DRracer) [inline lib/WUI/nhttp/gcode_upload.cpp:267]: :+1:

File: src/common/media.cpp [hunk 2]
```diff
@@ -183,6 +185,8 @@ char getByte(GCodeFilter::State *state) {
         media_loop_read++;
     } else if (feof(media_print_file)) {
         *state = GCodeFilter::State::Eof;
+    } else if (errno == EAGAIN) {
```
Comment (DRracer) [inline src/common/media.cpp]: :+1:

File: src/filesystem_fatfs.c [hunk 3]
```diff
@@ -92,7 +92,7 @@ static int get_errno(FRESULT result) {
     case FR_MKFS_ABORTED:
         return EIO;
     case FR_TIMEOUT:
-        return EIO;
+        return EAGAIN;
```
Comment (DRracer) [inline src/filesystem_fatfs.c:95]: :+1:

Comment (laloch) [reply src/filesystem_fatfs.c:95]: Yes, it is, but `fread`/`fwrite` is not required to set `errno` according to the POSIX standard anyway :wink:

File: lib/WUI/nhttp/gcode_upload.cpp [hunk 4]
```diff
@@ -118,6 +121,27 @@ GcodeUpload::UploadResult GcodeUpload::start(const RequestParser &parser, Upload
...
+        return StatusPage(Status::InsufficientStorage, StatusPage::CloseHandling::ErrorClose, json_errors, "USB drive full");
+    }
+    const size_t csize = svfs.f_frsize * svfs.f_bsize;
+    /* Pre-allocate the area needed for the file to prevent frequent metadata updates during the upload.
+       We do the pre-allocation incrementally cluster-by-cluster to minimize the chance of stalling
+       the media and blocking other tasks potentially waiting for the file system lock.
+       The file size doesn't need to be exactly the size of the uploaded content. We will truncate
+       the file later on. */
+    for (unsigned long sz = csize; sz < *parser.content_length; sz += csize) {
+        if (fseek(file.get(), sz, SEEK_SET)) {
```
Comment (DRracer) [inline lib/WUI/nhttp/gcode_upload.cpp:137]: Does `fseek` resize the file incrementally? Or is it some hidden "feature" of our implementation?

File: lib/WUI/nhttp/gcode_upload.cpp [hunk 5]
```diff
@@ -166,12 +190,17 @@ Step GcodeUpload::step(string_view input, bool terminated_by_client, uint8_t *,
...
-    const size_t written = fwrite(data.begin(), 1, data.size(), tmp_upload_file.get());
-    if (written < data.size()) {
-        // Data won't fit into the flash drive -> Insufficient stogare.
-        return make_tuple(Status::InsufficientStorage, "USB write error or USB full");
-    } else {
-        return make_tuple(Status::Ok, nullptr);
+    for (;;) {
+        const size_t written = fwrite(data.begin(), 1, data.size(), tmp_upload_file.get());
+        if (written < data.size()) {
+            if (errno == EAGAIN) {
```
Comment (DRracer) [inline lib/WUI/nhttp/gcode_upload.cpp:196]: :+1:

File: src/common/media.cpp [hunk 6]
```diff
@@ -76,12 +77,134 @@ char getByte(GCodeFilter::State *state);
 static char gcode_buffer[MAX_CMD_SIZE + 1]; // + 1 for NULL char
 static GCodeFilter gcode_filter(&getByte, gcode_buffer, sizeof(gcode_buffer));
 static uint32_t media_loop_read = 0;
-static const constexpr uint32_t MEDIA_LOOP_MAX_READ = 4096;
 
 media_state_t media_get_state(void) {
     return media_state;
 }
 
+#define PREFETCH_SIGNAL_START 1
```
Comment (DRracer) [inline src/common/media.cpp:85]: please prefer named constants (e.g. `static constexpr uint32_t something`) over preprocessor macros

File: src/common/media.cpp [hunk 7]
```diff
@@ -76,12 +77,134 @@ char getByte(GCodeFilter::State *state);
...
+#define PREFETCH_SIGNAL_STOP  2
+#define PREFETCH_SIGNAL_FETCH 4
+// These buffers are HUGE. We need to rework the prefetcher logic
+// to be more efficient and add compression.
+#define FILE_BUFF_SIZE 5120
+static char prefetch_buff[2][FILE_BUFF_SIZE];
```
Comment (DRracer) [inline src/common/media.cpp:91]: Are we sure that these buffers + functions are not used from anywhere else? E.g. PrusaLink trying to fetch thumbnails?

File: src/common/media.cpp [hunk 8]
```diff
@@ -76,12 +77,134 @@ char getByte(GCodeFilter::State *state);
...
+    file_buff = prefetch_buff[1];
+    for (;;) {
+        char *back_buff = prefetch_buff[0];
+        GCodeFilter::State bb_state = GCodeFilter::State::Ok;
+        uint32_t back_buff_level = 0;
+        osEvent event;
+
+        file_buff_level = file_buff_pos = 0;
+
+        event = osSignalWait(PREFETCH_SIGNAL_START | PREFETCH_SIGNAL_STOP | PREFETCH_SIGNAL_FETCH, osWaitForever);
```
Comment (DRracer) [inline src/common/media.cpp:111]: what happens if this loop really "waits forever"?

File: src/gui/screen_printing.cpp [hunk 9]
```diff
@@ -123,6 +123,8 @@ screen_printing_data_t::screen_printing_data_t()
     , w_time_value(this, Rect16(10, 148, 101, 20), is_multiline::no)
     , w_etime_label(this, Rect16(130, 128, 101, 20), is_multiline::no)
     , w_etime_value(this, Rect16(30, 148, 201, 20), is_multiline::no)
+    , g_queue_num(this, Rect16(10, 113, 100, 15), 0, " G-Queue %d", resource_font(IDR_FNT_SMALL))
```
Comment (DRracer) [inline src/gui/screen_printing.cpp:126]: visualization of read buffer level shall be refactored/improved in the future

File: src/usbh_diskio.c [hunk 10]
```diff
@@ -1,8 +1,8 @@
 #include "ff_gen_drv.h"
 #include "usbh_diskio.h"
-#include "timing.h"
 
 static const uint16_t USB_DEFAULT_BLOCK_SIZE = FF_MIN_SS;
+static DWORD scratch[FF_MAX_SS / 4];
```
Comment (DRracer) [inline src/usbh_diskio.c:5]: So in total this PR needs 2x 5KB + 1x 512B buffers, correct?

**Review by DRracer — APPROVED** [review]
Besides some minor comments this PR looks like a solid improvement of the dreadful USB reliability problem which was blocking us from releasing FW 4.4 with WiFi.

I can't verify the binary ESP blob in any other way than looking at a MINI fetching data from a web browser over WiFi, but that's good enough for now (i.e. finally bringing the WiFi feature to our users).

Numerous improvements can be done on top of this PR, especially:
- unit test the prefetch state machine + verify the edge cases (mutex waiting forever vs. some reasonable timeout)
- lower memory consumption (10.5KB feels quite large) - e.g. compress the data read from the G-code file seamlessly with minimal overhead
- if possible avoid doing memcpy - ideally let the DMA put data to the target memory area (it is not going to be a "nice" piece of code, but it can be done IMO)

File: src/common/media.cpp (hunk 6 above)
Comment (laloch) [reply src/common/media.cpp:85]: hmm, this should rather be an `enum` I guess, but OK, will do.

File: src/common/media.cpp (hunk 7 above)
Comment (laloch) [reply src/common/media.cpp:91]: I don't understand.

File: src/common/media.cpp (hunk 8 above)
Comment (laloch) [reply src/common/media.cpp:111]: Literally nothing. It's responsibility of the controlling thread to manage the state of this loop. FreeRTOS tasks are meant to never exit their loop. They are meant to either be terminated or run or be suspended forever.

File: src/gui/screen_printing.cpp (hunk 9 above)
Comment (laloch) [reply src/gui/screen_printing.cpp:126]: We don't want to keep this here. I don't think users care about these internals.

File: src/usbh_diskio.c (hunk 10 above)
Comment (laloch) [reply src/usbh_diskio.c:5]: Yes, plus the size of the prefetcher's stack. Either that or users of the FS need to align their buffers properly.

File: src/usbh_diskio.c [hunk 11]
```diff
@@ -58,8 +58,22 @@ DSTATUS USBH_status(BYTE lun) {
 DRESULT USBH_read(BYTE lun, BYTE *buff, DWORD sector, UINT count) {
     DRESULT res = RES_ERROR;
     MSC_LUNTypeDef info;
+    USBH_StatusTypeDef status = USBH_OK;
+
+    if ((DWORD)buff & 3) { // DMA Alignment issue, do single up to aligned buffer
+        while ((count--) && (status == USBH_OK)) {
+            status = USBH_MSC_Read(&hUsbHostHS, lun, sector + count, (uint8_t *)scratch, 1);
+            if (status == USBH_OK) {
+                memcpy(&buff[count * FF_MAX_SS], scratch, FF_MAX_SS);
```
[the remark this answers is not in the document]
Comment (laloch) [reply src/usbh_diskio.c:67]: No way! I'd rather turn the DMA back off than doing that.

File: lib/WUI/nhttp/gcode_upload.cpp (hunk 4 above)
Comment (laloch) [reply lib/WUI/nhttp/gcode_upload.cpp:137]: In general that depends on the FS driver implementation, but I can hardly imagine a FS driver that could do it atomically.

File: src/usbh_diskio.c (hunk 11 above)
[the remark this answers is not in the document]
Comment (laloch) [reply src/usbh_diskio.c:67]: Besides, that buffer is going to be used in a different way in the next revision.

[the remark this answers is not in the document]
Comment (DRracer) [reply src/usbh_diskio.c:67]: ok

**Review by DRracer — APPROVED** [review]

File: lib/WUI/nhttp/gcode_upload.cpp [hunk 12]
```diff
@@ -166,12 +190,17 @@ Step GcodeUpload::step(string_view input, bool terminated_by_client, uint8_t *,
...
+                continue;
+            }
+            // Data won't fit into the flash drive -> Insufficient stogare.
```
Comment (mkbel) [inline lib/WUI/nhttp/gcode_upload.cpp:199]: typo stogare
